js/ai/src/generate/action.ts

```diff
@@ -79,8 +81,14 @@ export const GenerateUtilParamSchema = z.object({
 export async function generateHelper(
   registry: Registry,
   input: z.infer<typeof GenerateUtilParamSchema>,
-  middleware?: ModelMiddleware[]
+  middleware?: ModelMiddleware[],
+  toolCallIteration?: number
 ): Promise<GenerateResponseData> {
+  toolCallIteration = toolCallIteration ?? 0;
+  const maxIterations = input.maxTurns ?? 5;
+  if (toolCallIteration > maxIterations) {
+    throw new Error(`Exceeded maximum tool call iterations (${maxIterations})`);
+  }
```

Comment:
Should be GenkitError including the response with `ABORTED` status

Reply:
Done.